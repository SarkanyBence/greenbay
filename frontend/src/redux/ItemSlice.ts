import {
  createSlice,
  PayloadAction,
  createAsyncThunk,
  SerializedError,
} from "@reduxjs/toolkit";
import fetchData from "../services/fetchData";
import ItemType from "../types/ItemType";

export interface ItemState {
  loading: boolean;
  items: ItemType[];
  error: SerializedError | null;
}
export const fetchAllItems = createAsyncThunk("/items", async (thunkApi) => {
  return await fetchData("GET", "/items");
});

export const ItemSlice = createSlice({
  name: "items",
  initialState: {
    loading: false,
    items: [] as ItemType[],
    error: null,
  },
  reducers: {
    addItem: (state: ItemState, action: PayloadAction<ItemType>) => {
      state.items = [...state.items, action.payload];
    },
    updateItem: (state: ItemState, action: PayloadAction<ItemType>) => {
      let filtered = state.items.filter(
        (item) => item.id !== action.payload.id
      );
      state.items = [...filtered, action.payload];
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchAllItems.pending, (state: ItemState) => {
      state.loading = true;
    });

    builder.addCase(fetchAllItems.fulfilled, (state: ItemState, action) => {
      state.loading = false;
      state.items = [...action.payload];
    });

    builder.addCase(fetchAllItems.rejected, (state: ItemState, action) => {
      state.error = action.error;
    });
  },
});

export const { addItem, updateItem } = ItemSlice.actions;

export const selectItems = (state: ItemState) => state;

export default ItemSlice.reducer;
