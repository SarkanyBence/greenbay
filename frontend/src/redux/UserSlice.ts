import { createSlice, PayloadAction } from "@reduxjs/toolkit";

export interface UserState {
  userName: string;
  isLoggedIn: boolean;
}

const initialState: UserState = {
  userName: "Guest",
  isLoggedIn: false,
};

export const userSlice = createSlice({
  name: "user",
  initialState: initialState,
  reducers: {
    changeUser: (state: UserState, action: PayloadAction<UserState>) => {
      state.userName = action.payload.userName;
      state.isLoggedIn = action.payload.isLoggedIn;
    },
    logoutUser: (state: UserState) => {
      state.userName = initialState.userName;
      state.isLoggedIn = initialState.isLoggedIn;
    },
  },
});

export const { changeUser, logoutUser } = userSlice.actions;

export const selectUser = (state: UserState) => state;
export const selectIsLoggedIn = (state: UserState) => state.isLoggedIn;
export const selectUserName = (state: UserState) => state.userName;

export default userSlice.reducer;
