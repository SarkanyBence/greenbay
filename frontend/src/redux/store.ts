import { configureStore } from "@reduxjs/toolkit";
import allReducers from "./combineReducers";

const store = configureStore({
  reducer: allReducers,
});

store.subscribe(() => {
  console.log("state updated", store.getState());
});

export default store;

export type StateType = ReturnType<typeof store.getState>;
export type DispatchType = typeof store.dispatch;
