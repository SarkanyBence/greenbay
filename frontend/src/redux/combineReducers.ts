import ItemSlice from "./ItemSlice";
import userSlice from "./UserSlice";

const allReducers = {
  user: userSlice,
  items: ItemSlice,
};

export default allReducers;
