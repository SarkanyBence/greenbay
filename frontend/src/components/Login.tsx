import { useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import { useAppDispatch } from "../hooks/stateHooks";
import fetchData from "../services/fetchData";
import NewUser from "../types/newUser";
import NewUserDto from "../types/newUserDto";
import PropsInput from "../types/PropsInput";
import User from "../types/User";
import FormInput from "./FormInput";
import { changeUser } from "../redux/UserSlice";
import { fetchAllItems } from "../redux/ItemSlice";

function Login() {
  const dispatch = useAppDispatch();
  const history = useHistory();
  const [userState, setUserState] = useState({
    userName: "",
    password: "",
  } as NewUser);

  const [userError, setUserError] = useState(false);

  useEffect(() => {}, [userError]);

  const submitForm = (e: any) => {
    e.preventDefault();
    const data = Object.fromEntries(new FormData(e.target).entries());
    const dataDto: NewUserDto = new NewUserDto(data as unknown as NewUser);

    fetchData("POST", "/login", dataDto)
      .then((token) => {
        localStorage.setItem("token", token);
        const newUser: User = {
          userName: userState.userName,
          isLoggedIn: true,
        };
        dispatch(changeUser(newUser));
        dispatch(fetchAllItems())
          .unwrap()
          .then((items) => {
            console.log(items);

            history.push("/buy");
          });
      })
      .catch((err) => {
        if (err.status === 400) {
          setUserError(true);
        }
      });
  };

  const handleChange = (e: any) => {
    setUserState({ ...userState, [e.target.name]: e.target.value });
  };

  const inputs: PropsInput[] = [
    {
      id: 1,
      name: "userName",
      placeholder: "Username",
      type: "text",
      errorMessage: "Username or password is incorrect",
      required: true,
    },
    {
      id: 2,
      name: "password",
      placeholder: "Password",
      type: "password",
      errorMessage: "Username or password is incorrect",
      required: true,
    },
  ];

  return (
    <div className="base-container">
      <form onSubmit={submitForm}>
        <span
          className="formInput other-error"
          style={{ opacity: userError ? 1 : 0 }}
        >
          Username or password is incorrect
        </span>
        {inputs.map((input) => (
          <FormInput
            key={input.id}
            {...input}
            userState={userState}
            handleChange={handleChange}
          />
        ))}

        <button>Login</button>
      </form>
    </div>
  );
}

export default Login;
