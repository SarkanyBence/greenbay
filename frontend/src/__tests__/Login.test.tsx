import Login from "../components/Login";
import * as ReactDom from "react-dom";
import renderer from "react-test-renderer";
import { UserState } from "../redux/UserSlice";
import * as reactRedux from "react-redux";
import { Provider } from "react-redux";
import store from "../redux/store";

describe("Login tests", () => {
  let container: HTMLDivElement;
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

  const initialState: UserState = {
    userName: "Guest",
    isLoggedIn: false,
  };

  const mockDispatch = jest.fn();
  jest.mock("react-redux", () => ({
    useAppSelector: jest.fn(),
    useAppDispatch: () => mockDispatch,
  }));

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
    container = document.createElement("div");
    document.body.appendChild(container);
    ReactDom.render(
      <Provider store={store}>
        <Login />
      </Provider>,
      container
    );
  });

  afterEach(() => {
    document.body.removeChild(container);
    container.remove();
  });

  it("Login snapshot test", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Login />{" "}
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.skip("renders correctly", () => {
    useSelectorMock.mockReturnValue({ initialState });
    const inputs = document.getElementsByClassName("input");
    expect(inputs.length).toBe(2);
  });
});
