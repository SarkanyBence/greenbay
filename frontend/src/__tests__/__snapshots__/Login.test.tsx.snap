// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests Login snapshot test 1`] = `
Array [
  <div
    className="base-container"
  >
    <form
      onSubmit={[Function]}
    >
      <span
        className="formInput other-error"
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        SOMETHING
      </span>
      <div
        className="formInput"
      >
        <input
          data-focused="false"
          name="userName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Username"
          required={true}
          type="text"
          value=""
        />
        <span
          className="formInput error"
        >
          Username or password is incorrect
        </span>
      </div>
      <div
        className="formInput"
      >
        <input
          data-focused="false"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
        <span
          className="formInput error"
        >
          Username or password is incorrect
        </span>
      </div>
      <button>
        Login
      </button>
    </form>
  </div>,
  " ",
]
`;
